bpf: Support connecting veth pairs to Linux routing

tgraf opened the pull request:
This PR contains the datapath changes required to support the plumbing model where a veth pair is hooked into Linux routing via individual routes or a bridge of some sort and where no `cilium_host` interface exists to act as a router between the host and endpoints.

This model will enable to provide chaining for:
 - aws-vpc-cni
 - Lyft CNI in veth mode
 - Openshift SDN
 - Calico
 - Any other CNI plugin that uses similar model

This is a dependency for #7862 

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7898)
<!-- Reviewable:end -->

tgraf committed:
bpf: Add to-container section to bpf_lxc

This makes the datapath logic previously available via cilium_host also
available via an egress hook on host facing veth peer. The section is
called to-container and will perform the following capabilities:

 * ingress policy
 * LB reverse NAT

This enables for the Linux routing layer to route directly into a veth pair or
similar which then forwards into another network namespace.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
loader: Allow to specify direction of BPF programs

This will be used to attach programs at egress on host facing veth pairs.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
loader: Support attaching program at egress for to-container section

When the endpoint configuration demands it, attach an egress bpf_lxc with
section "to-container"

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Enable ARP pass-through mode

Enable the datapath to pass through ARP requests when requested by the
endpoint's datapath configuration.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Skip ingress policy at egress of source if egress prog is in use

When a dedicated BPF prog is in use implementing ingress policy for an
endpoint, there is no need to enforce the ingress policy of the local
destination at the source. The policy will be enforced twice.

In this new mode, the identity of the source is carried in the skb mark to the
BPF program on the destination endpoint side. For this purpose, a new mark flag
is defined.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
api: Add EndpointDatapathConfiguration to PUT /endpoint/

Grants access to:
 - RequireARPPassthrough
 - RequireEgressProg
 - RequireRouting

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
endpoint: Do not release and restore IP for endpoint's with external IPAM

When the IP is managed externally, no release or re-allocation is required

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Allow to disable BPF based routing

This mode is useful if routing is entirely delegated to Linux routing.

Signed-off-by: Thomas Graf <thomas@cilium.io>

joestringer commented on test/bpf/check-complexity.sh:
```diff
@@ -36,8 +36,10 @@ function annotate_section_names {
-	    -e "s/\(section '2\/12'\)/\1 (tail_call IPV6_TO_LXC)/"
+	    -e "s/\(section '2\/11'\)/\1 (tail_call IPV4_TO_LXC_POLICY_ONLY)/" \
+	    -e "s/\(section '2\/12'\)/\1 (tail_call IPV6_TO_LXC_POLICY_ONLY)/" \
+	    -e "s/\(section '2\/13'\)/\1 (tail_call IPV4_TO_LXC)/" \
```
This is called `XXX_TO_ENDPOINT` above.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -955,6 +1033,66 @@ int tail_ipv4_policy(struct __sk_buff *skb)
+#endif
+
+	skb->cb[CB_SRC_LABEL] = 0;
+	ret = ipv4_policy(skb, 0, src_identity, &forwarding_reason, cfg);
```
How is L7 proxy handled here? If this is TC egress for veth to a container, it'll return `TC_ACT_OK` and pass through the container instead of redirecting to the proxy, right?

tgraf replied:
I've tested this before tproxy was merged. It is now broken. I'll have to see whether we can redirect to ingress of the host facing device

tgraf replied:
Fixed

joestringer reviewed (changes requested):
Initial glance, I plan to take a more thorough look in the coming days.

aanm commented on pkg/datapath/config.go:
```diff
@@ -54,4 +54,17 @@ type EndpointConfiguration interface {
+
+	// DisableRouting returns true if the endpoint requires for routing in
+	// BPF to be disabled and delegated to Linux routing
+	DisableRouting() bool
```
I would change the naming of this function as it seems the function itself will disable routing.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -822,6 +822,66 @@ int tail_ipv6_policy(struct __sk_buff *skb)
+	struct ipv6hdr *ip6;
+
+	if (!revalidate_data(skb, &data, &data_end, &ip6))
+		return DROP_INVALID;
```
No drop notifications here?

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -1043,8 +1163,83 @@ __section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_NAT46) int tail_ipv4_to_ipv6(struct
 #endif
 BPF_LICENSE("GPL");
+
+static inline bool __inline__ inherit_identity(struct __sk_buff *skb, __u32 *identity)
```
Can we share this implementation between `bpf_netdev` and `bpf_lxc`?

joestringer commented on api/v1/openapi.yaml:
```diff
@@ -946,6 +946,31 @@ definitions:
+          IP from being released and re-allocation of the IP address is skipped
+          on restore.
+        type: boolean
+      disable-routing:
```
Do we want to have negatively-defined options in the API?

tgraf replied:
Existing clients won't define this option, how else do you want to do it?

aanm replied:
@tgraf something like
```yaml
      enable-routing:
        description: >
          By enabling routing, all routing decisions are delegated to the
          cilium routing layer.
        type: boolean
        default: true
```

```golang
// RoutingEnabled returns false if...
func (e *Endpoint) RoutingEnabled() bool {
	return e.DatapathConfiguration.EnableRouting != nil && e.DatapathConfiguration.EnableRouting
}
```

tgraf replied:
I've been trying to avoid that to not introduce `*bool` but I'm OK with it. I agree that it is cleaner from an API perspective.

joestringer commented on daemon/state.go:
```diff
@@ -96,8 +96,11 @@ func (d *Daemon) validateEndpoint(ep *endpoint.Endpoint) (valid bool, err error)
-	if err := d.allocateIPsLocked(ep); err != nil {
-		return false, fmt.Errorf("Failed to re-allocate IP of endpoint: %s", err)
+	// If the IP is managed by an external IPAM, it does not need to be re-allocated
+	if !ep.DatapathConfiguration.ExternalIPAM {
```
The comment above this check and this check seem to be opposites?

I think the logic is correct, the comment just confused me.

joestringer reviewed (changes requested):
Some additional comments below.

tgraf committed:
endpoint: Add ability to install per endpoint route

Instead of routing everything through the cilium_host interface, allow the
endpoint to install a per endpoint route.

Signed-off-by: Thomas Graf <thomas@cilium.io>

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -299,6 +300,12 @@ type Endpoint struct {
+	// passed in via the plugin that created the endpoint, e.g. the CNI
+	// plugin which performed the plumbing will enable certain datapath
+	// features according to the mode selected.
+	DatapathConfiguration models.EndpointDatapathConfiguration
```
Maybe I missed this, but how is upgrade handled with respect to this configuration?

tgraf replied:
Right now, there are no plans to support upgrade. I've tried it and it is possible. Some of the required work was previously in this branch but this PR has already gotten big so I've removed it. As it stands now, whether to use per endpoint routes and direct linux routing is a decision made at cluster creation time, similar to IPVLAN or flannel mode.

Details:
Upgrade is possible but it requires new programs to keep a policy enforcement at egress as old endpoints may not protect themselves yet. I have a commit which does this via a flag that is enabled by default but disabled in new ConfigMaps. We could also consider automatically ripping that out once all old endpoints are gone. For now, I have chosen to reserve this discussion for a separate PR as it is not strictly required to establish the functionality of ENI routing and additional chaining.

joestringer replied:
I wasn't necessarily thinking of upgrade in terms of transitioning existing setups to the new model, just in terms of "is this populated correctly to achieve the old model when going from an existing setup" (minimally, try not to break things). I'm happy to push this out.

tgraf replied:
Oh, I see. All fields are opt-in except RequireRouting which defaults to true. The `RequireRouting()` on `Endpoint` will default to true when the field is not available.

ianvernon reviewed (approved):
Approving for API owner.
